File: scripts/models/combinator.lua
local constants = require "scripts.constants"
local PlayerSettings = require "scripts.models.player_settings"

---Converts any numeric value to a signed 32-bit integer.
---@param val any Number or string representation to convert.
---@return integer
local function ToInt32(val)
  local n = tonumber(val) or 0
  if n > 2147483647 then
    if n <= 4294967295 then
      return math.floor(n - 4294967296)
    else
      return 2147483647
    end
  elseif n < -2147483648 then
    return -2147483648
  end
  return math.floor(n)
end

---@class C2CC.NetworkSignal
---@field Signal SignalID
---@field Count integer

---@class C2CC.SectionGroupData
---@field public GroupIndex integer Section group index in control behavior.
---@field public GroupName string Formatted display group name.
---@field public RawGroupName string Raw unformatted group name string.
---@field public IsActive boolean Whether section is enabled/active.
---@field public MaxSlotFound integer Highest non-empty slot index.
---@field public Slots table<integer, { Signal: SignalID, Count: integer }> Slot filters data map.

---@class C2CC
---@field public Entity LuaEntity The Factorio combinator entity instance.
local C2CC = {}
C2CC.__index = C2CC

---@param entity LuaEntity
---@return C2CC
function C2CC:New(entity)
  local obj = setmetatable({}, C2CC)
  if entity and entity.valid then
    obj.Entity = entity
  end
  return obj
end

---@return LuaEntity|nil
function C2CC:GetEntity()
  if self.Entity and self.Entity.valid then
    return self.Entity
  end
  return nil
end

---@return LuaConstantCombinatorControlBehavior|nil
function C2CC:GetControlBehavior()
  local entity = self:GetEntity()
  if entity and entity.get_control_behavior then
    return entity.get_control_behavior()
  end
  return nil
end

---Ensures section at specified index exists on control behavior.
---@param cb LuaConstantCombinatorControlBehavior
---@param sectionIndex integer
---@return LuaLogisticSection|nil
local function getOrCreateSection(cb, sectionIndex)
  if not cb or not cb.valid then return nil end
  while cb.sections_count < sectionIndex do
    cb.add_section("")
  end
  return cb.get_section(sectionIndex)
end

---@param signal SignalID|nil
---@param count integer|string
---@return LogisticFilter|nil
local function MakeFilter(signal, count)
  if not signal or not signal.name then return nil end
  local minVal = ToInt32(count)
  local filter = {
    value = {
      type = signal.type or "item",
      name = signal.name,
      quality = signal.quality or "normal"
    },
    min = minVal
  }
  if signal.comparator then filter.comparator = signal.comparator end
  return filter
end

---@return boolean
function C2CC:IsEnabled()
  local cb = self:GetControlBehavior()
  if not cb then return true end
  return cb.enabled
end

---@param enabled boolean
function C2CC:SetEnabled(enabled)
  local cb = self:GetControlBehavior()
  if cb then
    cb.enabled = enabled
  end
end

---@return integer
function C2CC:GetPriority()
  local cb = self:GetControlBehavior()
  if not cb then return 0 end
  local sec = getOrCreateSection(cb, constants.SECTIONS.CYBERSYN_PRIORITY)
  if not sec then return 0 end
  local filter = sec.get_slot(1)
  if filter and filter.value and filter.value.name == constants.SETTINGS.CS_PRIORITY_NAME then
    return filter.min or 0
  end
  return 0
end

---@param value number|string
function C2CC:SetPriority(value)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = getOrCreateSection(cb, constants.SECTIONS.CYBERSYN_PRIORITY)
  if not sec then return end
  local n = ToInt32(value)
  if n ~= 0 then
    sec.set_slot(1, MakeFilter({ type = "virtual", name = constants.SETTINGS.CS_PRIORITY_NAME }, n))
  else
    sec.clear_slot(1)
  end
end

---@return C2CC.NetworkSignal|nil
function C2CC:GetNetworkSignal()
  local cb = self:GetControlBehavior()
  if not cb then return nil end
  local sec = getOrCreateSection(cb, constants.SECTIONS.NETWORK_MASK)
  if not sec then return nil end
  local filter = sec.get_slot(1)
  if filter and filter.value and filter.value.name then
    return { Signal = filter.value, Count = filter.min or 0 }
  end
  return nil
end

---@param signalData C2CC.NetworkSignal|nil
function C2CC:SetNetworkSignal(signalData)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = getOrCreateSection(cb, constants.SECTIONS.NETWORK_MASK)
  if not sec then return end
  if signalData and signalData.Signal and signalData.Signal.name then
    sec.set_slot(1, MakeFilter(signalData.Signal, signalData.Count))
  else
    sec.clear_slot(1)
  end
end

---@return C2CC.SectionGroupData[]
function C2CC:GetGroups()
  local cb = self:GetControlBehavior()
  if not cb then return {} end

  -- Ensure system sections (Section 1: Priority, Section 2: Network Mask) exist
  getOrCreateSection(cb, constants.SECTIONS.CYBERSYN_PRIORITY)
  getOrCreateSection(cb, constants.SECTIONS.NETWORK_MASK)

  local groups = {}
  for secIdx = 3, cb.sections_count do
    local sec = cb.get_section(secIdx)
    if sec and sec.valid then
      local rawName = sec.group or ""
      local displayName = (rawName ~= "") and rawName or "[no group assigned]"

      local slotMap = {}
      local maxSlotFound = 0
      local filtersCount = sec.filters_count or 0

      for i = 1, math.max(40, filtersCount) do
        local filter = sec.get_slot(i)
        if filter and filter.value and filter.value.name then
          slotMap[i] = {
            Signal = filter.value,
            Count = filter.min or 0
          }
          if i > maxSlotFound then
            maxSlotFound = i
          end
        end
      end

      table.insert(groups, {
        GroupIndex = sec.index,
        RawGroupName = rawName,
        GroupName = displayName,
        IsActive = sec.active ~= false,
        MaxSlotFound = maxSlotFound,
        Slots = slotMap
      })
    end
  end

  return groups
end

---@param groupIndex integer
---@param slotIndex integer
---@return SignalID|nil, integer
function C2CC:GetGroupSlot(groupIndex, slotIndex)
  local cb = self:GetControlBehavior()
  if not cb then return nil, 0 end
  local sec = cb.get_section(groupIndex)
  if not sec or not sec.valid then return nil, 0 end
  local filter = sec.get_slot(slotIndex)
  if filter and filter.value and filter.value.name then
    return filter.value, filter.min or 0
  end
  return nil, 0
end

---@param groupName string|nil
---@return LuaLogisticSection|nil
function C2CC:AddGroup(groupName)
  local cb = self:GetControlBehavior()
  if not cb then return nil end
  getOrCreateSection(cb, constants.SECTIONS.CYBERSYN_PRIORITY)
  getOrCreateSection(cb, constants.SECTIONS.NETWORK_MASK)
  return cb.add_section(groupName or "")
end

---@param groupIndex integer
---@param groupName string|nil
function C2CC:RenameGroup(groupIndex, groupName)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = cb.get_section(groupIndex)
  if sec and sec.valid then
    sec.group = groupName or ""
  end
end

---@param groupIndex integer
function C2CC:RemoveGroup(groupIndex)
  local cb = self:GetControlBehavior()
  if not cb then return end
  cb.remove_section(groupIndex)
end

---@param groupIndex integer
---@param isActive boolean
function C2CC:SetGroupActive(groupIndex, isActive)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = cb.get_section(groupIndex)
  if sec and sec.valid then
    sec.active = isActive ~= false
  end
end

---@param groupIndex integer
---@param slotIndex integer
---@param signal SignalID|nil
---@param count integer|string
function C2CC:SetGroupSlot(groupIndex, slotIndex, signal, count)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = cb.get_section(groupIndex)
  if not sec or not sec.valid then return end
  local filter = MakeFilter(signal, count)
  if filter then
    sec.set_slot(slotIndex, filter)
  else
    sec.clear_slot(slotIndex)
  end
end

---@param groupIndex integer
---@param slots table
function C2CC:SetGroupSlotsBulk(groupIndex, slots)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = cb.get_section(groupIndex)
  if not sec or not sec.valid then return end
  local filtersCount = sec.filters_count or 40
  for i = 1, math.max(40, filtersCount) do
    sec.clear_slot(i)
  end
  for k, v in pairs(slots or {}) do
    local sIdx = tonumber(k)
    if sIdx and v then
      local sig = v.Signal or v.signal
      local cnt = v.Count or v.count
      local filter = MakeFilter(sig, cnt)
      if filter then
        sec.set_slot(sIdx, filter)
      end
    end
  end
end

---@param groupIndex integer
---@param slotIndex integer
function C2CC:RemoveGroupSlot(groupIndex, slotIndex)
  local cb = self:GetControlBehavior()
  if not cb then return end
  local sec = cb.get_section(groupIndex)
  if sec and sec.valid then
    sec.clear_slot(slotIndex)
  end
end

---Copies settings, priority, network signal, and groups from another combinator instance.
---@param srcComb C2CC Source combinator instance.
function C2CC:CopyFrom(srcComb)
  if not srcComb then return end
  local destCb = self:GetControlBehavior()
  local srcCb = srcComb:GetControlBehavior()
  if not destCb or not srcCb then return end

  self:SetPriority(srcComb:GetPriority())
  self:SetNetworkSignal(srcComb:GetNetworkSignal())

  local destCsSec = getOrCreateSection(destCb, constants.SECTIONS.CYBERSYN_PRIORITY)
  local destNetSec = getOrCreateSection(destCb, constants.SECTIONS.NETWORK_MASK)

  for secIdx = destCb.sections_count, 3, -1 do
    destCb.remove_section(secIdx)
  end

  local srcGroups = srcComb:GetGroups()
  for _, grp in ipairs(srcGroups) do
    local newSec = self:AddGroup(grp.RawGroupName)
    if newSec and newSec.valid then
      newSec.active = grp.IsActive
      for slotIdx, sdata in pairs(grp.Slots) do
        if sdata and sdata.Signal then
          self:SetGroupSlot(newSec.index, slotIdx, sdata.Signal, sdata.Count)
        end
      end
    end
  end
end

---@param fromBlueprint boolean
---@param playerIndex? integer
function C2CC:InitializeDefaults(fromBlueprint, playerIndex)
  if fromBlueprint then return end
  local cb = self:GetControlBehavior()
  if not cb then return end

  local ps = PlayerSettings.Get(playerIndex or 1)

  local defaultPriority = tonumber(ps and ps.Priority) or constants.SETTINGS.DEFAULT_PRIORITY
  local defaultNetworkFlag = tonumber(ps and ps.NetworkFlag) or constants.SETTINGS.DEFAULT_NETWORK_FLAG
  local defaultNetworkSignal = (ps and ps.DefaultNetworkSignal) or constants.SETTINGS.DEFAULT_NETWORK_SIGNAL

  local s1 = getOrCreateSection(cb, constants.SECTIONS.CYBERSYN_PRIORITY)
  local s2 = getOrCreateSection(cb, constants.SECTIONS.NETWORK_MASK)

  if defaultPriority ~= 0 and s1 and not s1.get_slot(1).value then
    s1.set_slot(1, MakeFilter({ type = "virtual", name = constants.SETTINGS.CS_PRIORITY_NAME }, defaultPriority))
  end

  if defaultNetworkFlag ~= 0 and s2 and not s2.get_slot(1).value then
    s2.set_slot(1, MakeFilter(defaultNetworkSignal, defaultNetworkFlag))
  end

  -- Always initialize Section 3 (first custom section group) for new combinator placement
  if cb.sections_count < 3 then
    cb.add_section("")
  end
end

---Fixes sections after pasting settings from a vanilla/external constant combinator.
---Shifts pasted vanilla sections to Section 3+ and restores Section 1 (Priority) and Section 2 (Network Mask).
---@param playerIndex? integer Player index for default settings.
function C2CC:FixPastedVanillaSections(playerIndex)
  local cb = self:GetControlBehavior()
  if not cb then return end

  -- 1. Backup all current sections copied from the vanilla combinator
  local savedSections = {}
  for i = 1, cb.sections_count do
    local sec = cb.sections[i]
    if sec and sec.valid then
      local filters = {}
      for slotIdx = 1, sec.filters_count do
        local flt = sec.get_slot(slotIdx)
        if flt and flt.value then
          filters[slotIdx] = {
            value = flt.value,
            min = flt.min
          }
        end
      end
      table.insert(savedSections, {
        group_name = sec.group or "",
        active = sec.active,
        filters = filters
      })
    end
  end

  -- 2. Clear all sections
  for i = cb.sections_count, 1, -1 do
    cb.remove_section(i)
  end

  -- 3. Initialize Section 1 (Priority) and Section 2 (Network Mask) using player defaults
  self:InitializeDefaults(false, playerIndex)

  -- 4. Shift and re-create all vanilla sections into Section 3+
  for i, saved in ipairs(savedSections) do
    local targetSec = nil
    if i == 1 and cb.sections_count >= 3 then
      targetSec = cb.sections[3]
      targetSec.group = saved.group_name
    else
      targetSec = cb.add_section(saved.group_name)
    end

    if targetSec and targetSec.valid then
      targetSec.active = saved.active
      for slotIdx, flt in pairs(saved.filters) do
        targetSec.set_slot(slotIdx, {
          value = flt.value,
          min = flt.min
        })
      end
    end
  end
end

C2CC.ToInt32 = ToInt32

return C2CC

File: src/scripts/models/combinator.ts
import { SECTIONS, SETTINGS, ENTITY_NAME } from '../constants';
import { PlayerSettings } from './player_settings';
import { strace } from 'fcore/utils/strace';
import type {
  LuaConstantCombinatorControlBehavior,
  LuaLogisticSection,
  SignalID,
  LogisticFilter,
  LuaEntity,
  PlayerIndex,
  UnitNumber,
} from 'factorio:runtime';

export type Priority = number;
export type NetworkMask = number;
export type SectionIndex = number;
export type GroupIndex = number;
export type SlotIndex = number;

export function toInt32(val: any): number {
  if (val === undefined) return 0;
  const n = typeof val === 'number' ? val : tonumber(val as any);
  if (n === undefined || n !== n || n === Infinity || n === -Infinity) {
    return 0;
  }
  if (n > 2147483647) {
    if (n <= 4294967295) {
      return Math.floor(n - 4294967296);
    } else {
      return 2147483647;
    }
  } else if (n < -2147483648) {
    return -2147483648;
  }
  return Math.floor(n);
}

function getOrCreateSection(
  cb: LuaConstantCombinatorControlBehavior,
  sectionIndex: SectionIndex,
): LuaLogisticSection | undefined {
  if (!cb || !cb.valid) return undefined;
  while (cb.sections_count < sectionIndex) {
    cb.add_section('');
  }
  return cb.get_section(sectionIndex);
}

function makeFilter(
  signal: SignalID | undefined,
  count: number | string,
): LogisticFilter | undefined {
  if (!signal || !signal.name) return undefined;
  const minVal = toInt32(count);
  const filter: LogisticFilter = {
    value: {
      type: (signal.type || 'item') as any,
      name: signal.name,
      quality: (signal as any).quality || 'normal',
    },
    min: minVal,
  };
  if ((signal as any).comparator) {
    (filter as any).comparator = (signal as any).comparator;
  }
  return filter;
}

let nextSectionUid = 1;
const entitySectionUids: Record<UnitNumber, number[] | undefined> = {};

export class Combinator {
  public entity?: LuaEntity;

  constructor(entity?: LuaEntity) {
    if (entity && entity.valid) {
      this.entity = entity;
    }
  }

  public getEntity(): LuaEntity | undefined {
    return this.entity && this.entity.valid ? this.entity : undefined;
  }

  public getControlBehavior(): LuaConstantCombinatorControlBehavior | undefined {
    const ent = this.getEntity();
    if (ent && (ent as any).get_control_behavior) {
      return (ent as any).get_control_behavior() as LuaConstantCombinatorControlBehavior;
    }
    return undefined;
  }

  public isEnabled(): boolean {
    const cb = this.getControlBehavior();
    return cb ? cb.enabled : true;
  }

  public setEnabled(enabled: boolean): void {
    const cb = this.getControlBehavior();
    if (cb) {
      cb.enabled = enabled;
    }
  }

  public getPriority(): Priority {
    const cb = this.getControlBehavior();
    if (!cb) return 0;
    const sec = getOrCreateSection(cb, SECTIONS.CYBERSYN_PRIORITY);
    if (!sec) return 0;
    const filter = sec.get_slot(1);
    const val = filter?.value as any;
    if (val && (val.name === SETTINGS.CS_PRIORITY_NAME || val === SETTINGS.CS_PRIORITY_NAME)) {
      return filter.min || 0;
    }
    return 0;
  }

  public setPriority(value: Priority | string): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = getOrCreateSection(cb, SECTIONS.CYBERSYN_PRIORITY);
    if (!sec) return;
    const n = toInt32(value);
    if (n !== 0) {
      sec.set_slot(1, makeFilter({ type: 'virtual', name: SETTINGS.CS_PRIORITY_NAME }, n)!);
    } else {
      sec.clear_slot(1);
    }
  }

  public getNetworkSignal(): NetworkSignal | undefined {
    const cb = this.getControlBehavior();
    if (!cb) return undefined;
    const sec = getOrCreateSection(cb, SECTIONS.NETWORK_MASK);
    if (!sec) return undefined;
    const filter = sec.get_slot(1);
    const val = filter?.value as any;
    if (val && (val.name || typeof val === 'string')) {
      const sig: SignalID = typeof val === 'string' ? { name: val, type: 'item' } : val;
      return { signal: sig, count: filter.min || 0 };
    }
    return undefined;
  }

  public setNetworkSignal(netSig?: NetworkSignal): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = getOrCreateSection(cb, SECTIONS.NETWORK_MASK);
    if (!sec) return;
    if (netSig && netSig.signal) {
      sec.set_slot(1, makeFilter(netSig.signal, netSig.count || 0)!);
    } else {
      sec.clear_slot(1);
    }
  }

  public getGroups(): SectionGroupData[] {
    const cb = this.getControlBehavior();
    if (!cb) return [];
    getOrCreateSection(cb, SECTIONS.CYBERSYN_PRIORITY);
    getOrCreateSection(cb, SECTIONS.NETWORK_MASK);
    const groups: SectionGroupData[] = [];
    const count = cb.sections_count;
    const ent = this.getEntity();
    const unitNumber = ent?.unit_number;

    let uids: number[] | undefined;
    if (unitNumber) {
      uids = entitySectionUids[unitNumber];
      if (!uids) {
        uids = [];
        entitySectionUids[unitNumber] = uids;
      }
    }

    let arrIdx = 0;
    for (let i = 3; i <= count; i++) {
      const sec = cb.get_section(i);
      if (sec && sec.valid) {
        let uid = uids && arrIdx < uids.length ? uids[arrIdx] : undefined;
        if (uid === undefined && uids) {
          uid = nextSectionUid++;
          uids.push(uid);
        }
        arrIdx++;

        const slotMap: Record<SlotIndex, SectionGroupSlot | undefined> = {};
        let maxSlot = 0;
        const filtersCount = sec.filters_count || 40;
        for (let slotIdx = 1; slotIdx <= Math.max(40, filtersCount); slotIdx++) {
          const filter = sec.get_slot(slotIdx);
          const val = filter?.value as any;
          if (val && (val.name || typeof val === 'string')) {
            const sig: SignalID = typeof val === 'string' ? { name: val, type: 'item' } : val;
            slotMap[slotIdx] = {
              signal: sig,
              count: filter.min || 0,
            };
            if (slotIdx > maxSlot) maxSlot = slotIdx;
          }
        }

        const rawName = sec.group || '';
        groups.push({
          uid: uid,
          groupIndex: i,
          groupName: rawName || `Group ${i - 2}`,
          rawGroupName: rawName,
          isActive: sec.active !== false,
          maxSlotFound: maxSlot,
          slots: slotMap,
        });
      }
    }

    return groups;
  }

  public getGroupSlot(
    groupIndex: GroupIndex,
    slotIndex: SlotIndex,
  ): LuaMultiReturn<[SignalID | undefined, number]> {
    const cb = this.getControlBehavior();
    if (!cb) return $multi(undefined, 0);
    const sec = cb.get_section(groupIndex);
    if (!sec || !sec.valid) return $multi(undefined, 0);
    const filter = sec.get_slot(slotIndex);
    const val = filter?.value as any;
    if (val && (val.name || typeof val === 'string')) {
      const sig: SignalID = typeof val === 'string' ? { name: val, type: 'item' } : val;
      return $multi(sig, filter.min || 0);
    }
    return $multi(undefined, 0);
  }

  public addGroup(groupName?: string): LuaLogisticSection | undefined {
    const cb = this.getControlBehavior();
    if (!cb) return undefined;
    getOrCreateSection(cb, SECTIONS.CYBERSYN_PRIORITY);
    getOrCreateSection(cb, SECTIONS.NETWORK_MASK);
    const ent = this.getEntity();
    const unitNumber = ent?.unit_number;
    if (!unitNumber) return undefined;

    let uids = entitySectionUids[unitNumber];
    if (!uids) {
      uids = [];
      entitySectionUids[unitNumber] = uids;
    }
    uids.push(nextSectionUid++);
    return cb.add_section(groupName || '');
  }

  public renameGroup(groupIndex: GroupIndex, groupName?: string): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = cb.get_section(groupIndex);
    if (sec && sec.valid) {
      sec.group = groupName || '';
    }
  }

  public removeGroup(groupIndex: GroupIndex): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const ent = this.getEntity();
    const unitNumber = ent?.unit_number;
    if (!unitNumber) return;

    const uids = entitySectionUids[unitNumber];
    if (uids && groupIndex >= 3) {
      const arrIdx = groupIndex - 3;
      if (arrIdx < uids.length) {
        uids.splice(arrIdx, 1);
      }
    }
    cb.remove_section(groupIndex);
  }

  public setGroupActive(groupIndex: GroupIndex, isActive: boolean): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = cb.get_section(groupIndex);
    if (sec && sec.valid) {
      sec.active = isActive !== false;
    }
  }

  public setGroupSlot(
    groupIndex: GroupIndex,
    slotIndex: SlotIndex,
    signal?: SignalID,
    count: number | string = 0,
  ): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = cb.get_section(groupIndex);
    if (!sec || !sec.valid) return;
    const filter = makeFilter(signal, count);
    if (filter) {
      sec.set_slot(slotIndex, filter);
    } else {
      sec.clear_slot(slotIndex);
    }
  }

  public setGroupSlotsBulk(groupIndex: GroupIndex, slots: Record<SlotIndex | string, any>): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = cb.get_section(groupIndex);
    if (!sec || !sec.valid) return;
    const filtersCount = sec.filters_count || 40;
    for (let i = 1; i <= Math.max(40, filtersCount); i++) {
      sec.clear_slot(i);
    }
    for (const [k, v] of pairs(slots || {})) {
      const sIdx = tonumber(k);
      if (sIdx && v) {
        const sig = v.signal || v.Signal;
        const cnt = v.count || v.Count;
        const filter = makeFilter(sig, cnt);
        if (filter) {
          sec.set_slot(sIdx, filter);
        }
      }
    }
  }

  public removeGroupSlot(groupIndex: GroupIndex, slotIndex: SlotIndex): void {
    const cb = this.getControlBehavior();
    if (!cb) return;
    const sec = cb.get_section(groupIndex);
    if (sec && sec.valid) {
      sec.clear_slot(slotIndex);
    }
  }

  public copyFrom(srcComb: Combinator): void {
    if (!srcComb) return;
    const destCb = this.getControlBehavior();
    const srcCb = srcComb.getControlBehavior();
    if (!destCb || !srcCb) return;

    this.setPriority(srcComb.getPriority());
    this.setNetworkSignal(srcComb.getNetworkSignal());

    getOrCreateSection(destCb, SECTIONS.CYBERSYN_PRIORITY);
    getOrCreateSection(destCb, SECTIONS.NETWORK_MASK);

    for (let secIdx = destCb.sections_count; secIdx >= 3; secIdx--) {
      destCb.remove_section(secIdx);
    }

    const srcGroups = srcComb.getGroups();
    for (const grp of srcGroups) {
      const newSec = this.addGroup(grp.rawGroupName);
      if (newSec && newSec.valid) {
        newSec.active = grp.isActive;
        for (const [slotIdx, sdata] of pairs(grp.slots)) {
          if (sdata !== undefined && sdata.signal !== undefined) {
            this.setGroupSlot(newSec.index, tonumber(slotIdx) || 1, sdata.signal, sdata.count);
          }
        }
      }
    }
  }

  public initializeDefaults(fromBlueprint: boolean, playerIndex?: PlayerIndex): void {
    if (fromBlueprint) return;
    const cb = this.getControlBehavior();
    if (!cb) return;

    const ps = PlayerSettings.get((playerIndex || 1) as PlayerIndex);
    const defaultPriority = ps.priority;
    const defaultNetworkFlag = ps.networkFlag;
    const defaultNetworkSignal = ps.defaultNetworkSignal;

    const s1 = getOrCreateSection(cb, SECTIONS.CYBERSYN_PRIORITY);
    const s2 = getOrCreateSection(cb, SECTIONS.NETWORK_MASK);

    if (defaultPriority !== 0 && s1 && !s1.get_slot(1).value) {
      secSlotSet(
        s1,
        1,
        makeFilter({ type: 'virtual', name: SETTINGS.CS_PRIORITY_NAME }, defaultPriority)!,
      );
    }

    if (defaultNetworkFlag !== 0 && s2 && !s2.get_slot(1).value) {
      secSlotSet(s2, 1, makeFilter(defaultNetworkSignal, defaultNetworkFlag)!);
    }

    if (cb.sections_count < 3) {
      cb.add_section('');
    }
  }

  public fixPastedVanillaSections(playerIndex?: PlayerIndex): void {
    const cb = this.getControlBehavior();
    if (!cb) return;

    const savedSections: {
      groupName: string;
      active: boolean;
      filters: Record<number, LogisticFilter>;
    }[] = [];
    for (let i = 1; i <= cb.sections_count; i++) {
      const sec = cb.sections[i - 1];
      if (sec && sec.valid) {
        const filters: Record<number, LogisticFilter> = {};
        for (let slotIdx = 1; slotIdx <= sec.filters_count; slotIdx++) {
          const flt = sec.get_slot(slotIdx);
          if (flt && flt.value) {
            filters[slotIdx] = {
              value: flt.value,
              min: flt.min,
            };
          }
        }
        savedSections.push({
          groupName: sec.group || '',
          active: sec.active,
          filters: filters,
        });
      }
    }

    for (let i = cb.sections_count; i >= 1; i--) {
      cb.remove_section(i);
    }

    this.initializeDefaults(false, playerIndex);

    for (let i = 0; i < savedSections.length; i++) {
      const saved = savedSections[i];
      let targetSec: LuaLogisticSection | undefined;
      if (i === 0 && cb.sections_count >= 3) {
        targetSec = cb.sections[2];
        targetSec.group = saved.groupName;
      } else {
        targetSec = cb.add_section(saved.groupName);
      }

      if (targetSec && targetSec.valid) {
        targetSec.active = saved.active;
        for (const [slotIdx, flt] of pairs(saved.filters)) {
          targetSec.set_slot(tonumber(slotIdx) || 1, {
            value: flt.value,
            min: flt.min,
          });
        }
      }
    }
  }

  public static applyPriorityToAll(oldPriority: number, newPriority: number): number {
    if (!game) return 0;
    const oldPrio = tonumber(oldPriority) || 0;
    const newPrio = tonumber(newPriority) || 0;
    if (oldPrio === newPrio) return 0;

    let count = 0;
    for (const [_, surface] of pairs(game.surfaces)) {
      if (surface && surface.valid) {
        const combinators = surface.find_entities_filtered({ name: ENTITY_NAME });
        for (const entity of combinators) {
          if (entity && entity.valid) {
            const comb = new Combinator(entity);
            if (comb.getPriority() === oldPrio) {
              comb.setPriority(newPrio);
              count++;
            }
          }
        }
      }
    }
    return count;
  }

  public static applyNetworkToAll(
    oldNetworkSignal: SignalID | undefined,
    oldNetworkFlag: number | string,
    newNetworkSignal: SignalID | undefined,
    newNetworkFlag: number | string,
  ): number {
    if (!game) return 0;
    const oldFlag = tonumber(oldNetworkFlag) || 0;
    const newFlag = tonumber(newNetworkFlag) || 0;
    const sameSignal =
      oldNetworkSignal?.name === newNetworkSignal?.name &&
      (oldNetworkSignal?.type || 'item') === (newNetworkSignal?.type || 'item');
    if (sameSignal && oldFlag === newFlag) return 0;

    let count = 0;
    for (const [_, surface] of pairs(game.surfaces)) {
      if (surface && surface.valid) {
        const combinators = surface.find_entities_filtered({ name: ENTITY_NAME });
        for (const entity of combinators) {
          if (entity && entity.valid) {
            const comb = new Combinator(entity);
            const currentNet = comb.getNetworkSignal();
            let matchesOld = false;
            if (currentNet && currentNet.signal && oldNetworkSignal && oldNetworkSignal.name) {
              const sameType =
                (currentNet.signal.type || 'item') === (oldNetworkSignal.type || 'item');
              if (
                sameType &&
                currentNet.signal.name === oldNetworkSignal.name &&
                (tonumber(currentNet.count) || 0) === oldFlag
              ) {
                matchesOld = true;
              }
            }

            if (matchesOld) {
              if (newNetworkSignal && newNetworkSignal.name && newFlag !== 0) {
                comb.setNetworkSignal({ signal: newNetworkSignal, count: newFlag });
              } else {
                comb.setNetworkSignal(undefined);
              }
              count++;
            }
          }
        }
      }
    }
    return count;
  }
}

function secSlotSet(sec: LuaLogisticSection, slot: SlotIndex, filter: LogisticFilter) {
  sec.set_slot(slot, filter);
}

export const C2CC = Combinator;
